uvicorn/supervisors/multiprocess.py

```diff
@@ -32,39 +110,108 @@ def __init__(
+        self.signal_queue: list[int] = []
+        for sig in SIGNALS:
+            signal.signal(sig, lambda sig, frame: self.signal_queue.append(sig))
 
-        for _idx in range(self.config.workers):
-            process = get_subprocess(config=self.config, target=self.target, sockets=self.sockets)
+    def init_processes(self) -> None:
+        for _ in range(self.processes_num):
+            process = Process(self.config, self.target, self.sockets)
             process.start()
             self.processes.append(process)
 
-    def shutdown(self) -> None:
+    def terminate_all(self) -> None:
         for process in self.processes:
             process.terminate()
+
+    def join_all(self) -> None:
+        for process in self.processes:
             process.join()
 
-        message = f"Stopping parent process [{str(self.pid)}]"
-        color_message = "Stopping parent process [{}]".format(click.style(str(self.pid), fg="cyan", bold=True))
+    def restart_all(self) -> None:
+        for idx, process in enumerate(tuple(self.processes)):
+            process.terminate()
+            process.join()
+            new_process = Process(self.config, self.target, self.sockets)
+            new_process.start()
+            self.processes[idx] = new_process
+
+    def run(self) -> None:
+        message = f"Started parent process [{os.getpid()}]"
+        color_message = "Started parent process [{}]".format(click.style(str(os.getpid()), fg="cyan", bold=True))
+        logger.info(message, extra={"color_message": color_message})
+
+        self.init_processes()
+
+        while not self.should_exit.wait(0.5):
+            self.handle_signals()
+            self.keep_subprocess_alive()
+
+        self.terminate_all()
+        self.join_all()
+
+        message = f"Stopping parent process [{os.getpid()}]"
+        color_message = "Stopping parent process [{}]".format(click.style(str(os.getpid()), fg="cyan", bold=True))
         logger.info(message, extra={"color_message": color_message})
+
+    def keep_subprocess_alive(self) -> None:
+        for idx, process in enumerate(tuple(self.processes)):
+            if process.is_alive():
+                continue
+
+            process.kill()  # process is hung, kill it
+            process.join()
+            logger.info(f"Child process [{process.pid}] died")
+            del self.processes[idx]
+            process = Process(self.config, self.target, self.sockets)
+            process.start()
+            self.processes.append(process)
+
+    def handle_signals(self) -> None:
+        for sig in tuple(self.signal_queue):
+            self.signal_queue.remove(sig)
+            sig_name = SIGNALS[sig]
+            sig_handler = getattr(self, f"handle_{sig_name.lower()}", None)
+            if sig_handler is not None:
+                sig_handler()
+            else:  # pragma: no cover
+                logger.info(f"Received signal [{sig_name}], but nothing to do")
```

Comment:
I think `debug` is fine for _ignored_ signals.
```suggestion
                logger.debug(f"Received signal {sig_name}, but no handler is defined for it.")
```

Mainly suggesting this because of what I see locally when I resize my terminal window:
```bash
INFO:     Received signal [WINCH], but nothing to do
INFO:     Received signal [WINCH], but nothing to do
INFO:     Received signal [WINCH], but nothing to do
INFO:     Received signal [WINCH], but nothing to do
```